### PR TITLE
Do not do TransposeRewritePattern if the operation has more than one use

### PR DESCRIPTION
**Pull request by dhernandez0**
This PR checks that the transpose input has no other uses (apart from the transpose) before running the pass TransposeRewritePattern.

ticket: https://github.com/ROCm/rocMLIR-internal/issues/1563

### CODE REVIEW

**Comment by manupak** [comment]
The fix looks good. 

Please move e2e to test to where other such tests are.
Also please add a unit test in tosa-to-rock.mlir test file.

**Comment by manupak** [comment]
> ticket: [ROCm/rocMLIR-internal#1540](https://github.com/ROCm/rocMLIR-internal/issues/1540)

this is the wrong ticket.

File: mlir/test/Conversion/TosaToRock/tosa-to-rock.mlir [hunk 1]
```diff
@@ -25,3 +25,24 @@ func.func private @mlir_conv3d(%arg0: tensor<4x1x1x1x1xf32>, %arg1: tensor<2x5x5
...
+// CHECK-LABEL: mlir_dot_transpose_add
+// CHECK: %[[convRes:.*]] = rock.gemm %{{.*}} = %{{.*}} * %{{.*}} features =  none storeMethod =  set {arch = ""} : tensor<1x4x5xf32> = tensor<1x4x5xf32> * tensor<1x5x5xf32> -> tensor<1x4x5xf32>
+// CHECK-NEXT: %{{.*}} = tosa.reshape %[[convRes]]
+// CHECK-NEXT: %[[transRes:.*]] = rock.transform %[[convRes]] by #{{.*}} : tensor<1x4x5xf32> to tensor<1x5x4xf32>
+
+func.func private @mlir_dot_transpose_add(%arg0: tensor<20xf32>, %arg1: tensor<20xf32>, %arg2: tensor<25xf32>) -> (tensor<20xf32>, tensor<20xf32>) attributes {kernel, arch = ""} {
+  %0 = "tosa.reshape"(%arg0) {new_shape = array<i64: 1, 5, 4>} : (tensor<20xf32>) -> tensor<1x5x4xf32>
+  %1 = "tosa.reshape"(%arg2) {new_shape = array<i64: 1, 5, 5>} : (tensor<25xf32>) -> tensor<1x5x5xf32>
+  %2 = "tosa.reshape"(%arg1) {new_shape = array<i64: 1, 4, 5>} : (tensor<20xf32>) -> tensor<1x4x5xf32>
+  %3 = tosa.matmul %2, %1 : (tensor<1x4x5xf32>, tensor<1x5x5xf32>) -> tensor<1x4x5xf32>
```
Comment (manupak) [inline mlir/test/Conversion/TosaToRock/tosa-to-rock.mlir]: why this one different :) ?

File: mlir/test/Conversion/TosaToRock/tosa-to-rock.mlir [hunk 2]
```diff
@@ -25,3 +25,24 @@ func.func private @mlir_conv3d(%arg0: tensor<4x1x1x1x1xf32>, %arg1: tensor<2x5x5
...
+  %4 = "tosa.reshape"(%3) {new_shape = array<i64: 20>} : (tensor<1x4x5xf32>) -> tensor<20xf32>
+  %5 = "tosa.const"() <{value = dense<[0, 2, 1]> : tensor<3xi64>}> : () -> tensor<3xi64>
+  %6 = "tosa.transpose"(%3, %5) : (tensor<1x4x5xf32>, tensor<3xi64>) -> tensor<1x5x4xf32>
+  %7 = "tosa.add"(%6, %0) : (tensor<1x5x4xf32>, tensor<1x5x4xf32>) -> tensor<1x5x4xf32>
+  %8 = "tosa.reshape"(%7) {new_shape = array<i64: 20>} : (tensor<1x5x4xf32>) -> tensor<20xf32>
+  return %4, %8 : tensor<20xf32>, tensor<20xf32>
+}
```
Comment (manupak) [inline mlir/test/Conversion/TosaToRock/tosa-to-rock.mlir]: newline

**Review by manupak — COMMENTED** [review]
two nits

File: mlir/lib/Conversion/TosaToRock/TosaToRock.cpp [hunk 3]
```diff
@@ -632,6 +632,12 @@ struct TransposeRewritePattern : public OpRewritePattern<tosa::TransposeOp> {
     Value tInput = top.getOperand(0);
     Value tOutput = top.getResult();
 
+    // if the input has uses (apart from this one), we can't do this
+    if (!tInput.hasOneUse()) {
```
Comment (manupak) [inline mlir/lib/Conversion/TosaToRock/TosaToRock.cpp]: Ah this needs to be moved inside output fusion cases... i.e where tInput is checked to be a conv or gemm.
The third one -- `mergeTransposeWithGemmLikeOp` -- still could run.

**Review by manupak — APPROVED** [review]
LGTM! (you might get a codecov thing for conv as it not covered but I ll let you decide for now -- to fix that you might need add another case with conv in tosa-to-rock.mlir)

**Comment by krzysz00** [comment]
Nice catch on that bug from a few days ago (I think)

**Pull request merged**